## ğŸ§¬1. Genæ•°æ®å’ŒPheæ•°æ®

### åˆäºŒç”Ÿç‰©å­¦è¯¾æœ¬

![middle school](./images/middle.jpg)

### ğŸ“1.1 [HAPMAP3 genotype æ•°æ®](https://www.broadinstitute.org/medical-and-population-genetics/hapmap-3), 1ç™¾å¤šä¸‡ä¸ªSNPï¼Œä¸€èˆ¬ä½œä¸º LD è®¡ç®—çš„ reference panelã€‚

### ğŸ“1.2 [åƒäººåŸºå› ç»„é¡¹ç›®æ•°æ®](https://www.internationalgenome.org/data)ï¼Œ å°†è¿‘1äº¿ä¸ªSNPï¼Œä¸€èˆ¬ä½œä¸º imputation çš„ reference panelã€‚

### ğŸ“1.3 UKB æ•°æ®ï¼Œç°åœ¨æ¨èç”¨[UKB RAP](https://dnanexus.gitbook.io/uk-biobank-rap)ã€‚
![middle school](./images/ukb-disease.jpg)
```
1. æœ€æ–°æ•°æ®é€šçŸ¥ https://community.ukbiobank.ac.uk/hc/en-gb/articles/26088595922333-New-and-Updated-Data
2. UKB RAPï¼šhttps://ukbiobank.dnanexus.com/landing
```
<br/>


## ğŸ§¬2. GWAS

![GWAS](./images/GWAS.jpg)

### ğŸ“2.1 GWASæ•°æ®è·å–ï¼Œæœ€ç»å…¸çš„æ˜¯[GWAS Catalog](https://www.ebi.ac.uk/gwas)ã€‚

### ğŸ“2.2 GWASæ•°æ®QCç¤ºä¾‹ ã€æœ¬è¯¾é¢˜ç»„å»ºè®®GWASåˆ—åç§°ï¼š SNP CHR POS CHRPOS â­ EA NEA EAF N â­ BETA SE Z Pã€‘
```
1. ç¡®ä¿æ–‡ä»¶æ¯ä¸€è¡Œçš„åˆ—æ•°ç›®æ˜¯ä¸€æ ·çš„ã€‚å°†è¿ç»­ç©ºæ ¼ä¸­æ’å…¥NAï¼Œæ‰£å¥½ç¬¬ä¸€ç²’çº½æ‰£ã€‚
   zcat GWAS.gz | awk '{print NF}' | sort -nu | wc -l 
   sed 's/^\t/NA\t/; s/\t\t/\tNA\t/g; s/\t\t/\tNA\t/g; s/\t$/\tNA/'ã€‚
2. æŒ‰ç…§CHRå’ŒPOSæ’åºï¼Œå¦åˆ™pheweb ä¼šæŠ¥é”™
   sort -k 1,1V -k 2,2n # -Væ˜¯ä¸ºäº†æŠŠchrXå’ŒchrYæ’åˆ°æœ€åï¼Œä½†æ˜¯éœ€è¦æŠŠç¬¬ä¸€è¡Œå…ˆå†™åˆ°æ–°æ–‡ä»¶é‡Œã€‚
3. GWASæ•°æ®æœ¬èº«çš„é—®é¢˜ï¼š
   (1) Allele æœ€å¥½æ˜¯å¤§å†™ï¼Œawk å’Œ R éƒ½æœ‰ toupper()åŠŸèƒ½ã€‚
   (2) På€¼æœ€å¥½ä¸è¦å°äº1e-312ï¼Œawk ä¼šæŠŠå…¶å½“æˆ0ï¼Œæœ‰ä¸€äº›è½¯ä»¶ï¼ˆæ¯”å¦‚LDSCï¼‰ä¹Ÿä¼šæŠ¥é”™ï¼Œè¿™ä¸ªæ—¶å€™è¦ä¹ˆç”¨Zå€¼ï¼Œè¦ä¹ˆäººä¸ºå°†è¿™äº›På€¼è®¾ä¸º1e-300ã€‚
   (3) BETA|SE|På‡ºç°â€œä¸‰ç¼ºä¸€â€ çš„æƒ…å†µï¼š b = se * qnorm(p/2); se = abs(b/qnorm(p/2)); se = (CI_upper - CI_lower)/(1.96*2); p = 2*pnorm(-abs(b/se))

å¯¹æ£€æŸ¥æ²¡é—®é¢˜çš„GWASï¼Œæ·±åŠ å·¥ç¤ºä¾‹ï¼š

1. ğŸšœliftOver 
   dat=XYZ; head -1 $dat.txt > $dat.sorted; tail -n +2 $dat.txt | sort -k 1,1V -k 2,2n > $dat.sorted
   python ~/scripts/f/add_rsid.py -i $dat.sorted --sep "\t" --chr CHR --pos POS --ref NEA --alt EA -d ~/data/dbsnp/rsids-v154-hg19.tsv.gz -o $dat.tmp1
   cat $dat.tmp1 | awk 'NR >1 {print "chr"$1, $2 -1, $2, $9}' | sed 's/^chr23/chrX/' > $dat.tolift
   liftOver $dat.tolift /work/sph-huangj/files/hg19ToHg38.over.chain.gz $dat.lifted $dat.unmapped
   cut -f 3,4 $dat.lifted > $dat.pos_snp
   python ~/scripts/f/join_file.py -i "$dat.tmp1,TAB,8 $dat.pos_snp,TAB,1" -o $dat.tmp2
   cut -d " " -f 1-10 $dat.tmp2 | sed '1s/POS/POS.37/; 1s/NA/POS/' | gzip -f > clean/$dat.gz

2. â›„è·Ÿå…¶ä»–æ•°æ®åˆå¹¶ 
   python scripts/library/join_file.py -i "$dat,TAB,0 $dat.lifted.3col,TAB,2" -o $dat.NEW.tmp
   sed -i 's/  */\t/g' $dat.NEW.tmp; awk '$NF=="NA"' $dat.NEW.tmp | wc -l
   cut -f 1-10,12 $dat.NEW.tmp | sed '1 s/POS/POS.b38/' > $dat.NEW.txt

3. ğŸƒç˜¦èº« â€[æ¯”å¦‚è¯´FUMAä¸èƒ½æ¥å—è¶…è¿‡600MBçš„æ–‡ä»¶]
   zcat $dat.gz | awk 'function r(x) {return sprintf("%.4f", x)} {if (NR == 1) print; else if ($6 > 0.005 && $6 <= 0.995) {$6 = r($6); $8 = r($8); $9 = r($9); print}}' | sed 's/ /\t/g' | bgzip > $dat.lean.gz

4. ğŸ”ç´¢å¼• 
   tabix -f -S 1 -s 1 -b 2 -e 2 GWAS.gz
```

### ğŸ“2.3 GWASæ•°æ®å¯è§†åŒ–
>- å¯†è¥¿æ ¹å¤§å­¦å¼€å‘çš„[Pheweb](https://pheweb.org/)ï¼Œä¸Šé¢æ”¾äº†è‹±ç¾å¤§é˜Ÿåˆ—çš„æ•°åƒä¸ªGWASæ•°æ®ã€‚ æ­¤å¤–ï¼Œ[ä¸­å›½CKB](https://pheweb.ckbiobank.org/)ï¼Œ[ä¸­å›½å°æ¹¾TPMI](https://pheweb.ibms.sinica.edu.tw/)ï¼Œ[æ—¥æœ¬BBJ](https://pheweb.jp/)éƒ½ç”¨phewebå‘å¸ƒGWASã€‚
>- Phewebæœ‰ä¸€ä¸ªå¼ºå¤§çš„add_rsids.py çš„åŠŸèƒ½ï¼Œä½†æ˜¯å­˜åœ¨å…ˆå¤©ç¼ºé™·ï¼Œè§[èŠå¤©è®°å½•](https://github.com/statgen/pheweb/issues/217)ï¼Œç”¨æˆ·å¯ä»¥åœ¨å®‰è£…pheweb åæ‰¾åˆ° add_rsids.py æ–‡ä»¶ï¼ˆfind /home/ -name "add_rsid*" æˆ–è€… pip show --files phewebï¼‰ï¼Œä¿®æ”¹ä¸€è¡Œä»£ç ï¼ˆç¬¬140è¡Œï¼‰ã€‚
>- ç”¨æˆ·ä¹Ÿå¯ä»¥åœ¨[phewebèµ„æºåº“](https://resources.pheweb.org/)ç½‘ç«™ä¸‹è½½ rsids-v??-hg??.tsv.gz æ–‡ä»¶ï¼ˆ7äº¿å¤šè¡Œï¼‰ã€‚
>- å¦‚æœè¦ä»è¿™ä¸ªè¶…å¤§æ–‡ä»¶é‡Œæå–SNPçš„ä¿¡æ¯ï¼Œå¯ç”¨ bcftools view -i 'ID==@bmi.snp' rsids-v154-hg38.tsv.gz -Ou -o bmi.chrpos.txt
>- å¦‚æœGWASæ–‡ä»¶ â€œä¸‰ç¼ºä¸€â€ ï¼Œå¯ä»¥ä»scriptsæ–‡ä»¶å¤¹ä¸‹è½½æˆ‘æ”¹ç‰ˆçš„ snp_chrpos.pyï¼Œä¸€é”®è¡¥é½ï¼Œç¤ºä¾‹å‘½ä»¤å¦‚ä¸‹ã€‚ å¦‚æœæ²¡æœ‰ A1å’Œ A2åˆ—ï¼Œå°±ä¸ç”¨ --ref A1 --alt A2ã€‚
```
   python snp_chrpos.py -i bmi.gwas.gz --sep $'\t' --snp SNP --ref A1 --alt A2 -d data/dbsnp/rsids-v154-hg38.tsv.gz -o out.tsv
   python snp_chrpos.py -i bmi.gwas.gz --sep ',' --chr CHR --pos POS --ref NEA --alt EA -d data/dbsnp/rsids-v154-hg38.tsv.gz -o out.tsv
```

> å¯†è¥¿æ ¹å¤§å­¦è¿˜å¼€å‘äº†[locuszoom](http://locuszoom.org/) å®ç°åŸºå› ç»„å±€éƒ¨åœ°åŒºçš„å¯è§†åŒ–ğŸ”ã€‚ 
<br/>


## ğŸ§¬3. MR
>- å¦‚æœæœ‰ä¸ªä½“æ•°æ®ï¼Œå¯ä»¥ç”¨ [OneSampleMRåŒ…](https://cran.r-project.org/web/packages/OneSampleMR/index.html)ã€‚
>- å¦‚æœåªæœ‰å·²å‘è¡¨çš„summaryæ•°æ®ï¼Œå°±å¯ä»¥ä½¿ç”¨Bristolå¤§å­¦å¼€å‘çš„[TwoSampleMR RåŒ…](https://mrcieu.github.io/TwoSampleMR/index.html)æˆ–å‰‘æ¡¥å¤§å­¦å›¢é˜Ÿå¼€å‘çš„[MendelianRandomization RåŒ…](https://wellcomeopenresearch.org/articles/8-449)ã€‚
>- å·¥å…·å˜é‡ï¼Œä¸€èˆ¬éœ€è¦å»æ‰ F_stats <10 æˆ–è€…ä½äº <b>[MHCåŒºé—´]</b> ã€chr6:28477897-33448354 [(GRCh37)](https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC?asm=GRCh37), chr6:28510120-33480577 [(GRCh38)](https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC)ã€‘ çš„SNPã€‚
>- <b>å¯†è¥¿æ ¹å¤§å­¦</b>å¼€å‘çš„ [imputation server](https://imputationserver.sph.umich.edu) ç”¨çš„æ˜¯ï¼š ä»rs57232568 ã€29000554 (ç‰ˆæœ¬37), 29032777 (ç‰ˆæœ¬38)ã€‘ åˆ° rs116206961ã€33999992 (ç‰ˆæœ¬37), 34032215 (ç‰ˆæœ¬38)ã€‘
>- 10ä¸ªæ³¨æ„äº‹é¡¹ç¤ºä¾‹ï¼š
```
> 1. ç”¨ allele.qcï¼Œåè°ƒä¸¤ç»„æ•°æ®çš„ BETAå’ŒEAFï¼Œä½†æ˜¯è¾“å‡ºæ–‡ä»¶ä¾ç„¶æ˜¯åŸæ¥çš„EAå’Œ NEAã€‚
> 2. ç”¨ freadå’Œfwrite æ¯” read.table å’Œ write.table æ›´å¿«ï¼Œä½†fwriteé»˜è®¤è¾“å‡ºå¸¦quoteã€‚
> 3. EAFå¦‚æœæ˜¯0ï¼Œfwrite ä¼šå°†å…¶è§†ä¸ºNAï¼Œè¾“å‡ºåTXTå°±æ˜¯ä¸€ä¸ªç©ºæ ¼ï¼Œå¯¼è‡´æ•°æ®å°‘äº†ä¸€åˆ—ã€‚
> 4. gcta â€“cojo-slct ç”Ÿæˆçš„ .ldr.cojo æ–‡ä»¶æœ€åå¤šå‡ºä¸€åˆ—TABã€‚
> 5. æ·»åŠ ä»£ç ï¼Œå¤„ç†ä¸¤ä¸ªè¾“å…¥GWASï¼Œå…¶ä¸­ä¸€ä¸ªæˆ–è€…ä¸¤ä¸ªéƒ½ä¸å­˜åœ¨ EAF å’Œ N çš„é—®é¢˜ã€‚
> 6. è¿ç€ç”¨å‡ ä¸ª %>%ï¼Œç¼©å‡ä»£ç ï¼Œæœ€åå®é™…ä¸Šè·‘åäº†ã€å¤±æ§äº†ã€‚
> 7. ç”¨ group() ä¹‹åï¼Œå¦‚æœåé¢æ²¡æœ‰è·Ÿä¸Š ungroup()ï¼Œåé¢ä¼šå‡ºé—®é¢˜ã€‚
> 8. ç”¨ plinkï¼Œ.bim æ–‡ä»¶ä¸­çš„æŸ“è‰²ä½“å¯ä»¥ç”¨ Xè¡¨ç¤ºã€‚ä½†æ˜¯ gcta å¿…é¡»ç”¨ 23ã€‚
> 9. æ³¨æ„HLAçš„GRCh37æˆ–GRCh37çš„ç¡®åˆ‡ chr:start-endã€‚
>10. æœ€åå¯èƒ½å‘ç°è½¯ä»¶è·‘å‡ºæ¥çš„ç»“æœæœ‰é‡å¤§é—®é¢˜ï¼Œä¸â€œé²æ£’â€ https://github.com/ZhaotongL/cisMRcML/issues/6
```
<br/>


## ğŸ¤–4. AIç³»ç»Ÿ
>- Transformerï¼ˆæ¶æ„ï¼‰ï¼šæŒ‡ 2017 å¹´è®ºæ–‡ Attention Is All You Need æå‡ºçš„ç¥ç»ç½‘ç»œæ¶æ„ï¼Œæœ€åˆæ¥è‡ª Google å›¢é˜Ÿã€‚
>- Transformersï¼ˆåº“ï¼‰ï¼šæŒ‡ Hugging Face é‡Œçš„ Python åº“ï¼Œå¸¸æ­é… PyTorchï¼ˆtorchï¼‰ ä½¿ç”¨ï¼Œåšåº•å±‚è®­ç»ƒ/å¾®è°ƒã€‚
```
1.	æœ¬åœ°å®‰è£…å¤§æ¨¡å‹ï¼ˆä»¥åƒé—®ä¸ºä¾‹ï¼‰
	conda env list # conda env remove -n ai
	conda create -n ai python=3.11; conda activate ai
#	pip install --upgrade --no-cache-dir torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu128
	pip install --pre --upgrade --no-cache-dir torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu128
	pip install numpy tqdm transformers pandas requests openpyxl bitsandbytes
	pip install -U "transformers>=4.56.0" "accelerate>=1.10.1" datasets peft evaluate scikit-learn protobuf sentencepiece

2. 	pip install huggingface_hub; hf auth login; hf download google-bert/bert-base-chinese --local-dir . æˆ– git clone https://huggingface.co/Qwen/Qwen3-8B
	# å¦‚æœ Failed to connect to port 443ï¼Œå°±ç”¨ä¸‹é¢çš„pythonä»£ç ï¼š 
	import os, time
	from huggingface_hub import snapshot_download
	snapshot_download(repo_id="google-bert/bert-large-uncased", repo_type="model", local_dir="D:/data/ai/bert/bert-large-uncased")

3. å®‰è£… VS codeï¼Œåœ¨å·¦è¾¹Extensionsèœå•åˆ†åˆ«æœç´¢å¹¶å®‰è£… wslã€ pythonã€ jupyter
   wslé‡Œé¢ç”¨ which python, cmd é‡Œé¢ç”¨ where python, è€ŒVS code é‡Œé¢ç”¨ python -c "import sys; print(sys.executable)"
```
[![ç‚¹å‡»çœ‹è§†é¢‘](./images/nn-youtube.png)](https://www.youtube.com/playlist?list=PLZHQObOWTQDNU6R1_67000Dx_ZCJB-3pi)
<br/>

### å…³äºè›‹ç™½è´¨ç»“æ„é¢„æµ‹
>- ä¸‹è½½ [åƒäººåŸºå› ç»„æ·±åº¦æµ‹åºVCFæ–‡ä»¶](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20220422_3202_phased_SNV_INDEL_SV/), ä¸‹è½½[å‚è€ƒåŸºå› ç»„fastaæ–‡ä»¶](https://ftp.ensembl.org/pub/current_fasta/homo_sapiens/dna) ç„¶åsamtools faidx, 
ä»[gencode](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/) æˆ–è€… [ensembl](https://ftp.ensembl.org/pub/current_gtf/homo_sapiens/) ä¸‹è½½<b>gtf</b> å’Œ gff3æ–‡ä»¶, ä¸‹è½½[vcf2protè½¯ä»¶](https://github.com/ikmb/vcf2prot)ã€‚
>- åœ¨Alpha-foldæœåŠ¡å™¨è¾“å…¥çš„DNAï¼Œæ˜¯mRNAé€†è½¬å½•å½¢æˆçš„cDNAã€ä¸å«å†…å«å­ã€‘ï¼Œsnapgeneæˆ–Editseqå¯å°†DNAè½¬ä¸ºè›‹ç™½è´¨ğŸ¥šåºåˆ—
>- è›‹ç™½è´¨3Dä¹‹é—´ã€åŒ…æ‹¬å†·å†»ç”µé•œæ•°æ®ã€‘æ¯”è¾ƒï¼Œå¯ç”¨TMalign
>- å¾ˆå¤šæ ·æœ¬ _1 å’Œ _2 çš„è›‹ç™½å®Œå…¨ä¸€è‡´ï¼Œå…ˆ seqkit rmdup å»é‡ï¼Œå†è·‘ Alpha-Fold
```
bcftools +liftover chr9.vcf.gz -Oz -o chr9.lifted.vcf.gz -- -s $fasta_37 -f $fasta_38 -c $dir0/files/liftOver/hg19ToHg38.over.chain.gz 
bcftools view NEFL.csq.vcf -i 'INFO/BCSQ ~ "missense"' | bcftools query -f '[%SAMPLE\t%GT\n]' missense.vcf.gz | awk '$2 !~ /0\|0/' > missense.person
bcftools query ABO.csq.vcf.gz -f '%INFO/BCSQ\n' | tr ',' '\n' | awk -F'|' '{if ($3 ~ /^ENST/) print $3}' | sort -u
```
<br/>


## ğŸ§¬5. å‚è€ƒèµ„æ–™åŠç»éªŒåˆ†äº«

ğŸGWAS-PRS-MR â€ä¸‰é©¾é©¬è½¦â€œ å…¥é—¨æŒ‡å— 
``` 
> GWASå…¥é—¨ï¼š 2021. Nature RMP. [Genome-wide association studies](https://www.nature.com/articles/s43586-021-00056-9)  
> ğŸ®GWASè¯¦è§£ä¸­æ–‡ç‰ˆï¼šgwaslab.org
> PRSå…¥é—¨. Nature Protocols. [Tutorial: a guide to performing polygenic risk score analyses](https://www.nature.com/articles/s41596-020-0353-1)  
> MRå…¥é—¨ï¼š 2022. Nature RMP. [Mendelian randomization](https://www.nature.com/articles/s43586-021-00092-5)  
```

åŸºå› æ³¨é‡Šä¿¡æ¯ğŸ”
```
> dbSNP: https://www.ncbi.nlm.nih.gov/snp
> UCSC genome browser: https://www.genome.ucsc.edu
> ç¾å›½ç²¾å‡†åŒ»å­¦All of Usï¼š https://databrowser.researchallofus.org 
> å¯†è¥¿æ ¹å¤§å­¦å…¬å«å­¦é™¢ TopMed browser: https://bravo.sph.umich.edu
> ä¸€å¤©å‘äº†7ç¯‡ NATUREç³»åˆ—æ–‡ç« çš„Gnomadé¡¹ç›® browser: https://gnomad.broadinstitute.org
```

ğŸ›µR 
``` 
â–¸ WINDOWS â€œç¯å¢ƒå˜é‡â€é‡Œè®¾ç½®R_LIBS_USERï¼ŒLINUXåœ¨ ~/.Renvironè®¾ç½®ã€‚ ç”¨ .libPaths()æŸ¥çœ‹
â–¸ å…ˆå®‰è£… devtools, remotes åŒ…
â–¸ Rç”»å›¾é›†é”¦: https://r-graph-gallery.com/index.html  
â–¸ Ræ–°å† åœ°å›¾: https://statsandr.com/blog/top-r-resources-on-covid-19-coronavirus/  
â–¸ ä¾›å¤ç°ä»£ç ï¼š https://globalenvhealth.org/code-data-download/  
â–¸ ğŸ®é¡¾ç¥–å¹¿ç‚«é…·ç”Ÿä¿¡å›¾ï¼š [https://jokergoo.github.io/software/](https://jokergoo.github.io/software/)  
â–¸ ğŸ®æ¢å¿—ç”ŸRåŒ…èŸèƒ [https://gitee.com/sheng0825/projects](https://gitee.com/sheng0825/projects)  
```

ğŸ‡Ubuntu Linux æ“ä½œç³»ç»Ÿ
```
â­•Dç›˜çš„è·¯å¾„åˆ†åˆ«æ˜¯/mnt/dï¼Œä»¥æ­¤ç±»æ¨âš¡
> å½“æ‰“å¼€ shellï¼Œé‡åˆ°press any key to continueï¼Œç”¨ç®¡ç†å‘˜æƒé™æ‰“å¼€cmd, è¿è¡Œ netsh winsock reset
> åå°å¤šçº¿ç¨‹ä¸‹è½½: screen -dmS jack aria2c -x 4 -i url.txt --log-level=info --log=jack.log; screen -ls; screen -S jack -X quit 
> ä¸‰å‰‘å®¢ğŸ—¡ä»£ç ç¤ºä¾‹: awk '{cnt=int(NR/100); print $0 > "download"cnt".sh"}'
> HPC ç™»å½•ï¼š ssh sph-huangj@172.18.6.178 ã€å¤ªä¹™ã€‘ï¼› ssh -p 18188 sph-huangj@172.18.6.10 ã€å¯æ˜ã€‘
  åå°è¿è¡Œï¼š nohup ./assoc.sum.sh & ä¹‹å ps aux | grep ?.sh ä¹‹å kill
  ç¡¬ç›˜é¢åº¦ï¼šdu -h --max-depth=2; mmlsquota -g sph-huangj --block-size auto
  bsubç­‰: queueinfo -gpu -cpu; module avail  
```
>- åˆ›å›­301ğŸ–¨ï¼š ä»[å¯Œå£«å®˜ç½‘](https://m3support-fb.fujifilm-fb.com.cn/driver_downloads/www/)æœç´¢ ApeosPort C2060 ä¸‹è½½é©±åŠ¨ç¨‹åºï¼Œç„¶åè¿è¡Œã€‚ ğŸ‘‰â€œè®¾å¤‡ç±»å‹â€ é€‰TCP/IP ğŸ‘‰ æ‰“å°æœºIPä¸º 10.20.40.6
>- åˆ›å›­204ğŸ–¨ï¼šé¦–å…ˆè¿æ¥ LINK_7204æ— çº¿ç½‘ï¼Œå¯†ç æ˜¯???2025??04ï¼Œç„¶åä¸‹è½½[é©±åŠ¨ç¨‹åº](https://www.canon.com.cn/supports/download/simsdetail/0101228601.html?modelId=1524&channel=4)ï¼Œç‚¹å‡»ä¸€æ­¥æ­¥å®‰è£…ã€‚

ğŸŒ… ğŸŒ‡ ğŸŒ™ ğŸ¦Ÿ ğŸœ â–¸